File: mt-pricing-backend/app/schemas/human_queue.py
"""Schemas — Human Queue API (US-RND-01-10).

Incluye:
- ``HumanQueueItem``   — item de la cola (extiende MatchCandidateResponse).
- ``HumanQueueList``   — respuesta paginada del GET /human-queue.
- ``LabelRequest``     — body del POST /human-queue/{id}/label.
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

LabelLiteral = Literal["accept", "reject", "skip"]


class HumanQueueItem(BaseModel):
    """Item de la cola de validación humana."""

    model_config = ConfigDict(extra="ignore", from_attributes=True)

    id: UUID
    product_sku: str
    channel: str
    external_id: str
    brand: str | None = None
    title: str
    price_aed: Decimal | None = None
    specs_jsonb: dict = Field(default_factory=dict)
    kind: str = "unknown"
    score: int = Field(ge=0, le=100)
    status: str = "pending"
    calibrated_confidence: Decimal | None = None
    label: LabelLiteral | None = None
    reviewer_user_id: UUID | None = None
    reviewed_at: datetime | None = None
    validated_by: UUID | None = None
    validated_at: datetime | None = None
    discarded_reason: str | None = None
    created_at: datetime
    updated_at: datetime

    # VLM Judge (US-F15-02-02, AC#5/AC#6) — poblado desde specs_jsonb['vlm_judge'].
    # Nulo para rol viewer y cuando VLM no ha corrido.
    judge_rationale: str | None = None
    judge_image_regions: list[dict] | None = None


class HumanQueueList(BaseModel):
    """Respuesta paginada de GET /human-queue."""

    model_config = ConfigDict(extra="forbid")

    items: list[HumanQueueItem]
    total: int = Field(ge=0, description="Número de ítems en esta página")
    limit: int = Field(ge=1, le=200)
    offset: int = Field(ge=0)
    confidence_threshold: float = Field(ge=0.0, le=1.0)


class LabelRequest(BaseModel):
    """Body de POST /human-queue/{id}/label."""

    model_config = ConfigDict(extra="forbid")

    label: LabelLiteral = Field(description="Veredicto del revisor: accept / reject / skip")

File: mt-pricing-frontend/lib/api/endpoints/human-queue.ts
"use client";

import env from "@/lib/env";
import { createSupabaseBrowserClient } from "@/lib/supabase/client";

export type HumanQueueLabel = "accept" | "reject" | "skip";

export interface HumanQueueItem {
  id: string;
  product_sku: string;
  channel: string;
  external_id: string;
  brand: string | null;
  title: string;
  price_aed: string | null;
  specs_jsonb: Record<string, unknown>;
  kind: string;
  score: number;
  status: string;
  calibrated_confidence: string | null;
  label: HumanQueueLabel | null;
  reviewer_user_id: string | null;
  reviewed_at: string | null;
  validated_by: string | null;
  validated_at: string | null;
  discarded_reason: string | null;
  created_at: string;
  updated_at: string;
  // VLM Judge (US-F15-02-02, AC#5/AC#6) — null para viewers y cuando VLM no corrió
  judge_rationale: string | null;
  judge_image_regions: Record<string, string>[] | null;
}

export interface HumanQueueListResponse {
  items: HumanQueueItem[];
  total: number;
  limit: number;
  offset: number;
  confidence_threshold: number;
}

export interface HumanQueueFilters {
  limit?: number;
  offset?: number;
  confidence_threshold?: number;
}

export interface LabelPayload {
  label: HumanQueueLabel;
}

export class HumanQueueApiError extends Error {
  public readonly status: number;
  public readonly detail: unknown;

  constructor(status: number, detail: unknown, fallback: string) {
    super(typeof detail === "string" ? detail : fallback);
    this.name = "HumanQueueApiError";
    this.status = status;
    this.detail = detail;
  }
}

async function authedFetch<T>(path: string, init: RequestInit = {}): Promise<T> {
  const supabase = createSupabaseBrowserClient();
  const {
    data: { session },
  } = await supabase.auth.getSession();
  const headers = new Headers(init.headers);
  if (!headers.has("Content-Type") && init.body) {
    headers.set("Content-Type", "application/json");
  }
  if (session?.access_token) {
    headers.set("Authorization", `Bearer ${session.access_token}`);
  }
  const res = await fetch(`${env.NEXT_PUBLIC_BACKEND_URL}${path}`, {
    ...init,
    headers,
    cache: "no-store",
  });
  if (!res.ok) {
    let detail: unknown;
    try {
      detail = await res.json();
    } catch {
      /* noop */
    }
    throw new HumanQueueApiError(res.status, detail, res.statusText);
  }
  if (res.status === 204) return undefined as T;
  return (await res.json()) as T;
}

function buildQuery(filters: HumanQueueFilters): string {
  const params = new URLSearchParams();
  if (filters.limit !== undefined) params.set("limit", String(filters.limit));
  if (filters.offset !== undefined) params.set("offset", String(filters.offset));
  if (filters.confidence_threshold !== undefined) {
    params.set("confidence_threshold", String(filters.confidence_threshold));
  }
  const qs = params.toString();
  return qs ? `?${qs}` : "";
}

export const humanQueueApi = {
  list: (filters: HumanQueueFilters = {}): Promise<HumanQueueListResponse> =>
    authedFetch<HumanQueueListResponse>(
      `/api/v1/human-queue${buildQuery(filters)}`,
    ),

  label: (matchId: string, payload: LabelPayload): Promise<HumanQueueItem> =>
    authedFetch<HumanQueueItem>(
      `/api/v1/human-queue/${encodeURIComponent(matchId)}/label`,
      {
        method: "POST",
        body: JSON.stringify(payload),
      },
    ),
};
